File: modules/grayscale/android/src/main/java/expo/modules/grayscale/GrayscaleModule.kt
package expo.modules.grayscale

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.ColorMatrix
import android.graphics.ColorMatrixColorFilter
import android.graphics.Paint
import android.util.Base64
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.ByteArrayOutputStream

class GrayscaleModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("Grayscale")

    AsyncFunction("toGrayscale") { base64: String ->
      try {
        val result = convertToGrayscale(base64)
        mapOf("base64" to result)
      } catch (e: Exception) {
        mapOf("base64" to "")
      }
    }
  }

  private fun convertToGrayscale(base64Input: String): String {
    try {
      // Validate input
      if (base64Input.isEmpty()) {
        return ""
      }

      // Strip data URI prefix if present (e.g., "data:image/png;base64,")
      var base64String = base64Input
      if (base64String.contains("base64,")) {
        val components = base64String.split("base64,")
        if (components.size > 1) {
          base64String = components[1]
        }
      }

      // Remove all whitespace and newline characters
      base64String = base64String.replace("\\s".toRegex(), "")

      // Validate base64 string is not empty after cleaning
      if (base64String.isEmpty()) {
        return ""
      }

      // Decode base64 string to image data
      val decodedBytes = Base64.decode(base64String, Base64.DEFAULT)
      if (decodedBytes.isEmpty()) {
        return ""
      }

      // Decode with options to prevent OOM errors
      val options = BitmapFactory.Options().apply {
        inJustDecodeBounds = true
      }
      BitmapFactory.decodeByteArray(decodedBytes, 0, decodedBytes.size, options)

      // Check dimensions to prevent memory issues
      val maxDimension = 4096
      val width = options.outWidth
      val height = options.outHeight

      if (width <= 0 || height <= 0 || width > maxDimension || height > maxDimension) {
        return ""
      }

      // Calculate sample size if needed
      var sampleSize = 1
      if (width > maxDimension || height > maxDimension) {
        val widthRatio = (width.toFloat() / maxDimension).toInt()
        val heightRatio = (height.toFloat() / maxDimension).toInt()
        sampleSize = maxOf(widthRatio, heightRatio)
      }

      options.inJustDecodeBounds = false
      options.inSampleSize = sampleSize

      val originalBitmap = BitmapFactory.decodeByteArray(decodedBytes, 0, decodedBytes.size, options)
        ?: return ""

      // Convert to grayscale
      val grayscaleBitmap = convertImageToGrayscale(originalBitmap)
        ?: run {
          originalBitmap.recycle()
          return ""
        }

      // Convert to PNG and encode to base64
      val outputStream = ByteArrayOutputStream()
      grayscaleBitmap.compress(Bitmap.CompressFormat.PNG, 100, outputStream)
      val pngBytes = outputStream.toByteArray()

      // Clean up bitmaps
      if (originalBitmap != grayscaleBitmap) {
        originalBitmap.recycle()
      }
      grayscaleBitmap.recycle()

      return Base64.encodeToString(pngBytes, Base64.DEFAULT)
    } catch (e: Exception) {
      return ""
    }
  }

  private fun convertImageToGrayscale(originalBitmap: Bitmap): Bitmap? {
    try {
      // Create a new bitmap with the same dimensions
      val grayscaleBitmap = Bitmap.createBitmap(
        originalBitmap.width,
        originalBitmap.height,
        Bitmap.Config.ARGB_8888
      )

      // Create canvas to draw on the new bitmap
      val canvas = Canvas(grayscaleBitmap)

      // Create paint with grayscale color matrix
      val paint = Paint()
      val colorMatrix = ColorMatrix()
      colorMatrix.setSaturation(0f)
      val colorFilter = ColorMatrixColorFilter(colorMatrix)
      paint.colorFilter = colorFilter

      // Draw the original bitmap onto the new bitmap with grayscale filter
      canvas.drawBitmap(originalBitmap, 0f, 0f, paint)

      return grayscaleBitmap
    } catch (e: Exception) {
      return null
    }
  }
}

File: modules/grayscale/ios/GrayscaleModule.swift
import ExpoModulesCore
import UIKit

public class GrayscaleModule: Module {
  public func definition() -> ModuleDefinition {
    Name("Grayscale")

    AsyncFunction("toGrayscale") { (base64: String) -> [String: String] in
      return autoreleasepool {
        let result = self.convertToGrayscale(base64: base64)
        return ["base64": result]
      }
    }
  }
  
  private func convertToGrayscale(base64: String) -> String {
    // Validate input
    guard !base64.isEmpty else {
      return ""
    }
    
    // Strip data URI prefix if present (e.g., "data:image/png;base64,")
    var base64String = base64
    if base64.contains("base64,") {
      let components = base64.components(separatedBy: "base64,")
      if components.count > 1 {
        base64String = components[1]
      }
    }
    
    // Remove all whitespace and newline characters
    base64String = base64String.components(separatedBy: .whitespacesAndNewlines).joined()
    
    // Validate base64 string is not empty after cleaning
    guard !base64String.isEmpty else {
      return ""
    }
    
    // Decode base64 string to image data
    guard let originalImageData = Data(base64Encoded: base64String, options: .ignoreUnknownCharacters),
          !originalImageData.isEmpty,
          let originalImage = UIImage(data: originalImageData) else {
      return ""
    }
    
    // Validate image size to prevent memory issues
    let maxDimension: CGFloat = 4096
    let imageSize = originalImage.size
    if imageSize.width > maxDimension || imageSize.height > maxDimension {
      // Scale down if too large
      let scale = min(maxDimension / imageSize.width, maxDimension / imageSize.height)
      let scaledSize = CGSize(width: imageSize.width * scale, height: imageSize.height * scale)
      
      guard let scaledImage = self.scaleImage(image: originalImage, toSize: scaledSize) else {
        return ""
      }
      
      guard let grayscaleImage = self.convertImageToGrayscale(image: scaledImage) else {
        return ""
      }
      
      guard let pngData = grayscaleImage.pngData() else {
        return ""
      }
      
      return pngData.base64EncodedString(options: .lineLength64Characters)
    }
    
    // Convert to grayscale
    guard let grayscaleImage = self.convertImageToGrayscale(image: originalImage) else {
      return ""
    }
    
    // Convert to PNG and encode to base64
    guard let pngData = grayscaleImage.pngData() else {
      return ""
    }
    
    return pngData.base64EncodedString(options: .lineLength64Characters)
  }
  
  private func scaleImage(image: UIImage, toSize size: CGSize) -> UIImage? {
    autoreleasepool {
      UIGraphicsBeginImageContextWithOptions(size, false, image.scale)
      defer { UIGraphicsEndImageContext() }
      
      image.draw(in: CGRect(origin: .zero, size: size))
      return UIGraphicsGetImageFromCurrentImageContext()
    }
  }
  
  private func convertImageToGrayscale(image: UIImage) -> UIImage? {
    return autoreleasepool {
      // Create image rectangle with current image width/height
      let imageRect = CGRect(x: 0, y: 0, width: image.size.width, height: image.size.height)
      
      // Grayscale color space
      guard let colorSpace = CGColorSpace(name: CGColorSpace.genericGrayGamma2_2) else {
        return nil
      }
      
      // Validate dimensions to prevent memory issues
      let width = Int(image.size.width)
      let height = Int(image.size.height)
      
      guard width > 0 && height > 0 && width <= 4096 && height <= 4096 else {
        return nil
      }
      
      // Create bitmap context with current image size and grayscale colorspace
      guard let context = CGContext(
        data: nil,
        width: width,
        height: height,
        bitsPerComponent: 8,
        bytesPerRow: 0,
        space: colorSpace,
        bitmapInfo: CGImageAlphaInfo.none.rawValue
      ) else {
        return nil
      }
      
      // Draw image into current context, with specified rectangle
      // using previously defined context (with grayscale colorspace)
      guard let cgImage = image.cgImage else {
        return nil
      }
      context.draw(cgImage, in: imageRect)
      
      // Create bitmap image info from pixel data in current context
      guard let grayscaleCGImage = context.makeImage() else {
        return nil
      }
      
      // Create a new UIImage object
      let newImage = UIImage(cgImage: grayscaleCGImage)
      
      return newImage
    }
  }
}
